cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ClusterCa.java

```diff
@@ -177,22 +177,14 @@ public Map<String, CertAndKey> generateZkCerts(String namespace, String kafkaNam
 
     public Map<String, CertAndKey> generateBrokerCerts(String namespace, String cluster, int replicas, Set<String> externalBootstrapAddresses,
                                                        Map<Integer, Set<String>> externalAddresses, boolean isMaintenanceTimeWindowsSatisfied) throws IOException {
-        DnsNameGenerator kafkaDnsGenerator = DnsNameGenerator.of(namespace, KafkaResources.bootstrapServiceName(cluster));
-        DnsNameGenerator kafkaHeadlessDnsGenerator = DnsNameGenerator.of(namespace, KafkaResources.brokersServiceName(cluster));
-
         Function<Integer, Subject> subjectFn = i -> {
             Subject.Builder subject = new Subject.Builder()
                     .withOrganizationName("io.strimzi")
                     .withCommonName(KafkaResources.kafkaStatefulSetName(cluster));
 
-            subject.addDnsName(KafkaResources.bootstrapServiceName(cluster));
-            subject.addDnsName(String.format("%s.%s", KafkaResources.bootstrapServiceName(cluster), namespace));
-            subject.addDnsName(kafkaDnsGenerator.serviceDnsNameWithoutClusterDomain());
-            subject.addDnsName(kafkaDnsGenerator.serviceDnsName());
-            subject.addDnsName(KafkaResources.brokersServiceName(cluster));
-            subject.addDnsName(String.format("%s.%s", KafkaResources.brokersServiceName(cluster), namespace));
-            subject.addDnsName(kafkaHeadlessDnsGenerator.serviceDnsNameWithoutClusterDomain());
-            subject.addDnsName(kafkaHeadlessDnsGenerator.serviceDnsName());
+            subject.addDnsNames(ModelUtils.generateAllServiceDnsNames(namespace, KafkaResources.bootstrapServiceName(cluster)));
```

Comment:
in case of clusterIp listener, I think this list is already filled in externalBootstrapAddresses

Reply:
The certificate exists once per broker regardless the number of listeners. In this place, it adds the dns names for the default services. That is needed regardless any user-configured-listeners since it is used in any case for replication and control-plane listener. So this handles the default services. So for example with cluster named `my-cluster`, this will add DNS names for services `my-cluster-kafka-brokers` and `my-cluster-kafka-bootstrap`.

The place in the KafkaListenerReconciler handles the DNS names for the listeners. When I add a `type: cluster-ip` listener named `tls`, it would for example add the DNS names for the bootstrap service `my-cluster-kafka-tls-bootstrap`. So these will be different from those added in the `ClusterCA` class.